LRA TCK refactoring while adding Arquillian as integration point for the test

ochaloup opened the pull request:
fixes #94

Refactoring of the TCK tests which provides Arquillian to base the integration on.
This adds Arquillian handling, changing few names and adding MicroProfile Config to the play.

When agreed on this approach and this is about to be merged I would like to continue with some more refactoring of the TCK codebase. There are still some comments and tests which needs to be clarified. This mostly migrates tests to be run with Arquillian but it does not do any changes in the logic of the tests.

The concept of the Arquillian integration is based on other spec (https://github.com/eclipse/microprofile-fault-tolerance/blob/master/tck/running_the_tck.asciidoc, https://github.com/eclipse/microprofile-health/blob/master/tck/running_the_tck.asciidoc).
The changes for Narayana LRA client are at: https://github.com/ochaloup/narayana/tree/lra-tck-arquillian-rework

As MicroProfile Config is used the timeout is configurable with changes provided here. The https://github.com/eclipse/microprofile-lra/issues/60 are about to be covered by this PR.

Can you please take a look how you feel about this?:  @mmusgrov @rdebusscher @xstefank

xstefank commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -600,137 +664,130 @@ private void joinAndEnd(boolean waitForRecovery, boolean close, String path, Str
-            if (readEntity) {
-                return response.readEntity(String.class);
-            }
+            assertEquals("Not expected status at call ' " + resourcePath.getUri() + "'",
```
redundant space after '

xstefank commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/participant/activity/ActivityStorage.java:
```diff
@@ -17,35 +17,42 @@
+    public Activity getActivityAndAssertExistence(String lraId, UriInfo jaxrsContext) {
+        if(!activities.containsKey(lraId)) {
+           String errorMessage = String.format("Activity store does not contain LRA id '%s' "
+                   + "while it was invoked method at '%s'", lraId, jaxrsContext.getPath());
```
invoked by the method at?

xstefank commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/participant/api/LraController.java:
```diff
@@ -485,19 +478,19 @@ public Response timeLimit(@HeaderParam(LRA_HTTP_HEADER) String lraId) {
             Thread.sleep(200);
         } catch (InterruptedException e) {
-            e.printStackTrace();
+            LOGGER.log(Level.FINE, "Interrupted because the revewed time limit elapsed", e);
```
renewed

xstefank commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/participant/api/Util.java:
```diff
@@ -142,6 +146,62 @@ public static boolean isAsyncCompletion(Method method) {
+        if(value < 0){
+            throw new IllegalArgumentException("value to adjust can't be negative");
+        }
+        return (int) Math.ceil(value * factor);
```
you can reuse long variant I think

ochaloup commented on tck/running_the_tck.asciidoc:
```diff
@@ -14,38 +15,119 @@
+  time that the tests waits for the coordinator start-up
+`lra.tck.coordinator.hostname`::
+  hostname where coordinator is available at for TCK suite can run the LRA status from the coordinator
+`lra.tck.coordinator.port`::
```
correct, it should be possible and it should not be ambiguous as currently is. I will check how is that possible.

ochaloup commented:
@xstefank 
> +100 for adding javadoc explaining test scenarios
I see your point and I'm really for adding the javadoc comment to describe what is the purpose of the testcases. Just bear in mind that the current change is meant to re-factor from old http execution style to junit+arquillian. Comments were not part of the codebase before and they are not either for now.

I didn't implement the test cases. If I add the comments I need to study the purpose and reason about validity of the test cases. Please, believe me, I can promise I will add javadoc description as the next step. 
Just for now, I would really like this re-factoring being committed first.

ochaloup commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -448,148 +530,130 @@ private String leaveLRAViaAPI() throws WebApplicationException {
      * Participants can pass data during enlistment and this data will be returned during
      * the complete/compensate callbacks
      */
+    // TODO: is this test to be run?
```
cleaning tests are planned as the next step. I need to discuss the intent with the author.

mmusgrov replied:
Ditto (I forget why I disabled some tests but the ones you have marked @Ignore will need adding to issue #92 )

mmusgrov commented on tck/running_the_tck.asciidoc:
```diff
@@ -14,38 +15,91 @@
+  When set-up lower then the timeouts will be shorter.
+`lra.tck.coordinator.waiting`::
+  Time that the tests waits for the LRA recovery starts-up.
+`lra.http.host`, `lra.http.port`::
```
Could we use an authority strings for endpoints instead (https://tools.ietf.org/html/rfc3986#section-3.2) or maybe just say it is a URI which gives us maximum flexibility.

ochaloup replied:
this parameters will be removed completely as discussed at the meeting the coordinator should not be expected by TCKs

mmusgrov commented on tck/running_the_tck.asciidoc:
```diff
@@ -14,38 +15,91 @@
-To enable the tests in your implementation of this specification you need to add the
-following dependency to your build:
+* `pom.xml` dependencies are set-up
+* MicroProfile Config is providedby runtime for the LRA TCK suite could be run
```
This doesn't read very well. Instead I would just say:

> MicroProfile Config

Or:

> the LRA TCK suite requires MicroProfile Config

ochaloup replied:
thanks!

mmusgrov commented on tck/running_the_tck.asciidoc:
```diff
@@ -14,38 +15,91 @@
+* a default Arquillian container is configured in `arquillian.xml` (tests manually deploy with use of `@ArquillianResource Deployer`)
+* implementation provides one implementation of `org.eclipse.microprofile.lra.client.LRAClient` as it's injected by TCK suite
+* implementation has to provide one implementation of `org.eclipse.microprofile.lra.tck.spi.ManagementSPI`. This is an interface
+  with definition of util methods used by the TCK suite for its run.
```
I guess you will remove this requirement if we end up removing the ManagementSPI interface.

ochaloup replied:
yes, correct

mmusgrov commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -45,398 +36,508 @@
+     */
+    @Inject @ConfigProperty(name = LRAClient.LRA_RECOVERY_PATH_KEY, defaultValue = "lra-recovery-coordinator")
+    private int recoveryPath;
+
```
PR #68 is removing all of these properties

ochaloup replied:
ok, I see. There will be just `lra.http.recovery.url`. When the PR#68 is merged I can change this for url too.

mmusgrov commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -45,398 +36,508 @@
-    }
-
-    public TckResult runTck(String testname, boolean verbose) {
-        TckResult run = new TckResult();
```
Excellent, it's good that we no longer need to run the TCK in this manner.

mmusgrov commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -45,398 +36,508 @@
-        lraClient.close();
-        msClient.close();
-        rcClient.close();
+        if(tckSuiteClient != null) tckSuiteClient.close();
```
Do we use checkstyle in the TCK?

ochaloup replied:
yes, the checkstyle is used. You mean that you don't like this without use of curly braces? 

```
if(tckSuiteClient != null)  {
    tckSuiteClient.close();
}
```

If you are about to add this rule to the checkstyle I can change the code here.

mmusgrov replied:
I think that all projects should be using the same set of rules but I see that ours is different since we have commented out NeedBraces:

>  <!--module name="NeedBraces" / -->

I would recommend that that rule is re-enabled.

I will check to see where else we are different from other projects and if we are I will raise a separate issue for it (hopefully this is the only one).

mmusgrov commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -45,398 +36,508 @@
     @Test
-    private String isCompensatedLRA() throws WebApplicationException {
-        URL lra = lraClient.startLRA(null, "SpecTest#isCompensatedLRA", LRA_TIMEOUT_MILLIS, ChronoUnit.MILLIS);
+    @Ignore
```
I forget why I disabled some tests but the ones you have marked @Ignore will need adding to issue #92

ochaloup replied:
+1

yes, the tests marked as `@Ignore` are not run in current TCK. Even they were not marked as Ignore they were rejected from running by not being defined in the list of permitted tests.
Are you fine that this step of checking what are purposes of tests will be run as separate follow-up task after this PR will be merged and when SmallRye client implementation is merged to repo (for being possible to run the TCK tests - e.g. those ignored ones - with some existing implementation)?

mmusgrov commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -45,398 +36,508 @@
     @Test
-    private String isCompletedLRA() throws WebApplicationException {
-        URL lra = lraClient.startLRA(null, "SpecTest#isCompletedLRA", LRA_TIMEOUT_MILLIS, ChronoUnit.MILLIS);
+    @Ignore
```
Ditto (I forget why I disabled some tests but the ones you have marked @Ignore will need adding to issue #92 )

ochaloup replied:
+1

mmusgrov commented on tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java:
```diff
@@ -46,398 +36,490 @@
+     */
     @Test
-    private String joinLRAViaBody() throws WebApplicationException {
+    @Ignore
```
Ditto (I forget why I disabled some tests but the ones you have marked @Ignore will need adding to issue #92 )

ochaloup commented on tck/running_the_tck.asciidoc:
```diff
@@ -14,38 +15,91 @@
+  The URL where the TCK suite deployment is exposed at. The TCK suite will construct path based on this URL.
+  The default base URL is `http://localhost:8180`.
+`lra.tck.timeout.factor`::
+  Timeout factor of default value `1.0`. When set bigger than `1.0` then timeout value will be bigger and waiting time are longer.
```
the `lra.tck.timeout.factor` is used to influence any timeout (any waiting time) in the TCK suite I enhanced the description to be more comprehensible.
I removed the waiting time completely for now. The TCK implementor can handle the waiting time for coordinator startup on his own.

ochaloup commented:
> It's looking great, I only had a few comments.
Thanks. I fixed the point you have. I would like to remove the ManagementSPI  in the next step when each test case should be examined.

> Also can you indicate somewhere how to debug the tests (either in the docs or on this PR).
I added a short section on debugging but the way depends on the runtime where the TCK suite is deployed.

mmusgrov commented:
@ochaloup Have you tried testing the new TCK against any LRA implementations yet?

ochaloup commented:
@mmusgrov as mentioned in the description I was working with Narayana implementation https://github.com/ochaloup/narayana/tree/lra-tck-arquillian-rework. But I think it does not follow all the changes happened during last week (or two). The branch needs changes to get it aligned.

mmusgrov commented:
Can we control which snapshot build get's pulled in?

ochaloup commented:
@mmusgrov for Narayana I don't know other way then define the snapshot in pom.xml. Or we can be creating some tags in this repo and get those particular ones built in CI.
But I think the best would be to just fix the Narayana LRA when some change happens in the spec.

ochaloup commented:
@mmusgrov I updated the checkstyle as you mentioned and fixed the merge trouble caused by the changes in spec

rdebusscher commented:
> @mmusgrov as mentioned in the description I was working with Narayana implementation https://github.com/ochaloup/narayana/tree/lra-tck-arquillian-rework. But I think it does not follow all the changes happened during last week (or two). The LRA client branch needs changes to be aligned and working.

@ochaloup 

Does the TCK run with implementation and coordinator? It should not be all green and use our latest decisions, but we need to make sure that it runs. Finetuning (on the tck side and the implementation) can be done afterward.

ochaloup commented:
I just fixed typos that was mentioned by @xstefank in his comments and I forgot to address them. Otherwise no other changes with this force push.

ochaloup commented:
@rdebusscher the idea is that the implementator of the TCK will ensure:

- to start and running the coordinator (or the coordination endpoints defined by the spec)
- configure the Arquillian to run with his implementation for Arquillian can start a container and deploy there tests
- to implement the prescribed API/SPI (currently it's `ManagementSPI`)

So the Narayana TCK runner is about to start the Narayana LRA coordinator and to configure Arquillian to be capable to deploy tests.

Is that what you meant?

rdebusscher commented:
@ochaloup 
>Is that what you meant?
Yes and no :)

We just need to make sure this updated code actually run with 'something' already (since the TCK is part of something bigger and cannot run on his own. You need those 3 points you mentioned)
But if this runs with the Narayana TCK, then it is fine.
